extension/parquet/parquet_crypto.cpp

```diff
@@ -116,47 +110,50 @@ class EncryptionTransport : public TTransport {
 		// Write length
 		const auto ciphertext_length = allocator.SizeInBytes();
 		const uint32_t total_length = ParquetCrypto::NONCE_BYTES + ciphertext_length + ParquetCrypto::TAG_BYTES;
-		trans.write(const_data_ptr_cast(&total_length), ParquetCrypto::LENGTH_BYTES);
 
-		// Write nonce
+		trans.write(const_data_ptr_cast(&total_length), ParquetCrypto::LENGTH_BYTES);
+		// Write nonce at beginning of encrypted chunk
 		trans.write(nonce, ParquetCrypto::NONCE_BYTES);
 
-		// Encrypt and write data
 		data_t aes_buffer[ParquetCrypto::CRYPTO_BLOCK_SIZE];
 		auto current = allocator.GetTail();
+
+		// Loop through the whole chunk
 		while (current != nullptr) {
 			for (idx_t pos = 0; pos < current->current_position; pos += ParquetCrypto::CRYPTO_BLOCK_SIZE) {
 				auto next = MinValue<idx_t>(current->current_position - pos, ParquetCrypto::CRYPTO_BLOCK_SIZE);
 				auto write_size =
-				    aes.Process(current->data.get() + pos, next, aes_buffer, ParquetCrypto::CRYPTO_BLOCK_SIZE);
+				    aes->Process(current->data.get() + pos, next, aes_buffer, ParquetCrypto::CRYPTO_BLOCK_SIZE);
 				trans.write(aes_buffer, write_size);
 			}
 			current = current->prev;
 		}
 
-		// Finalize the last encrypted data and write tag
+		// Finalize the last encrypted data
 		data_t tag[ParquetCrypto::TAG_BYTES];
-		auto write_size = aes.Finalize(aes_buffer, ParquetCrypto::CRYPTO_BLOCK_SIZE, tag, ParquetCrypto::TAG_BYTES);
+		auto write_size = aes->Finalize(aes_buffer, 0, tag, ParquetCrypto::TAG_BYTES);
 		trans.write(aes_buffer, write_size);
+		// Write tag for verification
 		trans.write(tag, ParquetCrypto::TAG_BYTES);
 
 		return ParquetCrypto::LENGTH_BYTES + total_length;
 	}
 
 private:
-	void Initialize() {
-		// Generate nonce and initialize AES
-		GenerateNonce(nonce);
-		aes.InitializeEncryption(nonce, ParquetCrypto::NONCE_BYTES);
+	void Initialize(const string &key) {
+		// Generate Nonce
+		aes->GenerateRandomData(nonce, ParquetCrypto::NONCE_BYTES);
+		// Initialize Encryption
+		aes->InitializeEncryption(nonce, ParquetCrypto::NONCE_BYTES, &key);
 	}
 
 private:
 	//! Protocol and corresponding transport that we're wrapping
 	TProtocol &prot;
 	TTransport &trans;
 
-	//! AES context
-	AESGCMState aes;
+	//! AES context and buffers
+	shared_ptr<AESGCMState> aes;
```

Comment:
I realize this was here before as well - but Parquet does support [different encryption algorithms](https://github.com/apache/parquet-format/blob/master/Encryption.md#41-aes-modes-used-in-parquet). Perhaps we can call this encryption data as this might be extended in the future?

Reply:
The different encryption algorithms by Parquet are  AES GCM and a combination of AES GCM and AES CTR, where in the latter case only data pages itself are encrypted with CTR but the other modules with GCM. Considering that they only support AES, we can maybe just rename it to AESState?